Improve compatibility with non-whole-minute timezone offsets

Algunenano opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve compatibility with non-whole-minute timezone offsets


Detailed description / Documentation draft:
DateLUT was assuming timezone offsets are whole minutes but that's not true, for example Liberia used to be `-0:43:08` and later became `-0:44:30` (it currently is 00:00, but that's less exciting).

Closes https://github.com/ClickHouse/ClickHouse/issues/27079
It appeared as part of the CI (in https://github.com/ClickHouse/ClickHouse/pull/26978)

Algunenano committed:
Improve compatibility with non-whole-minute timezone offsets

Algunenano committed:
Make gcc happy

Algunenano commented:
2 things in CI:
* Multiple checksum reports (https://clickhouse-test-reports.s3.yandex.net/27080/f2f778376295c077224fcd6f9242f6f0164a4e5f/functional_stateful_tests_(address).html#fail1): it looks like a hardware issue as I've built ASAN locally and run those tests without issues. The odd thing is that they are connection to localhost, maybe memory?
* PVS check. I'll try to fix it

Let's see if a new commit fixing (2) also shows the address issues or it has been something temporary.

Algunenano committed:
Linter

amosbird commented:
> for example Liberia used to be `-0:43:08`

That's interesting...

Algunenano commented:
Rebased with master. No new changes (it was clean already)

alexey-milovidov commented:
We have 2x performance degradation of `toSecond` function or somewhat similar:
https://clickhouse-test-reports.s3.yandex.net/27080/1b58e276b8b0534e925c79e9764896c847f9ec7f/performance_comparison/report.html#fail1

alexey-milovidov commented on base/common/DateLUTImpl.h:
```diff
@@ -474,38 +493,23 @@ class DateLUTImpl
 
-        return time / 60 % 60;
+        Time res = time / 60 % 60;
+        if (likely(res >= 0))
```
How the opposite is possible if we are dividing the number of seconds after midnight (always positive)?

Algunenano replied:
I was probably thinking about negative values of `t` (with Datetime64), which will lead to negative `time`  and negative `res`; but that doesn't see to be supported anyway.

alexey-milovidov commented on base/common/DateLUTImpl.h:
```diff
@@ -459,7 +464,21 @@ class DateLUTImpl
     inline unsigned toSecond(Time t) const
     {
-        auto res = t % 60;
+        if (offset_is_whole_number_of_hours_during_epoch)
```
likely?

Algunenano commented:
> We have 2x performance degradation of toSecond function or somewhat similar:

It might be because `Asia/Kolkata` is at `UTC+5:30` and that won't use the fast path (`offset_is_whole_number_of_hours_during_epoch`). I'll see if there is something better to do in this situation,

alexey-milovidov commented:
I'm not sure about this change.
We did not support non whole 15 minutes offsets - this is deliberate choice.

... but if we can avoid any performance degradation.

Algunenano commented:
I think the performance degradation can be solved in a similar way as the whole hour thing, by adding another flag like `offset_is_whole_number_of_minutes_during_epoch`. It's not great, but it should limit the performance impact to the offsets were the responses were wrong. I'll give it a try (likely on Monday).

alexey-milovidov commented:
Ok, sounds good.

Algunenano committed:
Improvements based on review

Algunenano commented:
Rebased and applied the improvements from the comments.

Performance of the affected query:
* Initially (before changes): `QPS: 12.465, RPS: 623806343.190, MiB/s: 4759.265, result RPS: 12.465`
* After initial changes: `QPS: 3.516, RPS: 175981641.031, MiB/s: 1342.633, result RPS: 3.516`
* Now: `QPS: 12.487, RPS: 624930109.263, MiB/s: 4767.838, result RPS: 12.487`

Looking at how things work internally (DateTimeTransforms.h), it looks like the fast path (common offsets) of some of the functions (toSecond, toMinute) could be vectorized using the same architecture as other functions (accepting an array instead of a single element), but that's something for a different time.

alexey-milovidov commented:
Ok.

Theoretically it can also be auto-vectorized, because the function is inlined. But I doubt that compiler is smart enough to perform "unlikely" branch checks in batches...

Algunenano committed:
Revert toMinute changes and add additional roundDown fastpath

Algunenano commented:
I added an invalid optimization (ok for minutes, not for hours). I'm looking into it and will try to unify calls/code if possible. Sadly, changing DateLutImpl.h requires recompiling ~1300 files so it takes time to verify and benchmark

Algunenano committed:
Separate minute intervals

Instead of mixing multiple rounding, use toStartOfMinuteInterval for all
minute related calculations

alexey-milovidov commented:
> GithubException: 500 null

Looks like GitHub API is down.

alexey-milovidov commented:
@mergifyio update

mergify[bot] committed:
Merge branch 'master' into timezone_00719

alexey-milovidov commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into timezone_00719

alexey-milovidov reviewed (approved):
Wow, everything is alright :)
